src/engine/GroupBy.cpp

```diff
@@ -759,57 +760,162 @@ GroupBy::checkIfHashMapOptimizationPossible(
+                     T&& singleResult) mutable {
+    auto generator = sparqlExpression::detail::makeGenerator(
+        std::forward<T>(singleResult),
+        evaluationContext._endIndex - evaluationContext._beginIndex,
+        &evaluationContext);
+
+    // TODO<C++23> use views::enumerate
+    auto idx = 0;
+    for (sparqlExpression::IdOrString val : generator) {
+      (*resultTable)(evaluationContext._beginIndex + idx++, outCol) =
+          sparqlExpression::detail::constantExpressionResultToId(std::move(val),
+                                                                 *localVocab);
+    }
+  };
+
+  std::visit(visitor, std::forward<sparqlExpression::ExpressionResult>(
+                          expressionResult));
+}
+
+// _____________________________________________________________________________
+template <size_t numAggregates>
+sparqlExpression::VectorWithMemoryLimit<ValueId>
+GroupBy::extractValuesDirectlyFromMap(
+    sparqlExpression::EvaluationContext& evaluationContext,
+    const ad_utility::HashMapWithMemoryLimit<KeyType, ValueType<numAggregates>>&
+        map,
+    IdTable* resultTable, size_t hashMapIndex, size_t outCol) {
+  // Store in a vector such that we can copy it to
+  // `_previousResultsFromSameGroup`.
+  sparqlExpression::VectorWithMemoryLimit<ValueId> aggregateResults(
+      getExecutionContext()->getAllocator());
+
+  // Store aggregate results in table and in vector.
+  for (size_t j = evaluationContext._beginIndex;
+       j < evaluationContext._endIndex; j++) {
+    auto& val = (*resultTable)(j, 0);
+    auto& aggregateData = map.at(val).at(hashMapIndex);
+    auto aggregateResult = aggregateData.calculateResult();
+    (*resultTable)(j, outCol) = aggregateResult;
+    aggregateResults.push_back(aggregateResult);
+  }
+
+  return aggregateResults;
+}
+
+// _____________________________________________________________________________
+template <size_t numAggregates>
+void GroupBy::substituteAllAggregates(
+    std::vector<HashMapAggregateInformation>& info,
+    sparqlExpression::EvaluationContext& evaluationContext,
+    const ad_utility::HashMapWithMemoryLimit<KeyType, ValueType<numAggregates>>&
+        map,
+    IdTable* groupValues) {
+  // Multiple aggregates, need to substitute them away and then finally
+  // evaluate the top-level expression.
+  for (auto& aggregate : info) {
+    sparqlExpression::VectorWithMemoryLimit<ValueId> aggregateResults(
+        getExecutionContext()->getAllocator());
+
+    // Get all aggregate results as a vector
+    for (size_t j = evaluationContext._beginIndex;
+         j < evaluationContext._endIndex; j++) {
+      auto& val = (*groupValues)(j, 0);
+      auto& aggregateData = map.at(val).at(aggregate.hashMapIndex_);
+      aggregateResults.push_back(aggregateData.calculateResult());
+    }
+
+    // Substitute the resulting vector as a literal
+    std::unique_ptr<sparqlExpression::SparqlExpression> newExpression =
```

Comment:
use `auto newExpression = make_unique<>` The type is already clear enough by the `make_unique`.